File: Python/backend/utils/Logger.py
"""
Logger.py — regista eventos em dois destinos em paralelo:
  1. ficheiros .txt em logs/<ano>/logs<Tipo><mes>.txt
  2. tabela `logs` na base de dados SQLite
"""

import os
from datetime import datetime

from ..models.Marcacao import Marcacao
from ..models.Pendente import Pendente
from . import Database


def _escrever_log(tipo_txt: str, tipo_bd: str, mensagem: str):
    """
    tipo_txt : sufixo do ficheiro
    tipo_bd  : valor na coluna tipo da BD
    mensagem : texto do log
    """
    now = datetime.now()
    data = now.strftime("%d/%m/%Y %H:%M:%S")
    ano = now.strftime("%Y")
    mes = now.strftime("%m")
    linha = f"[{data}] {mensagem}\n"

    # ficheiro .txt
    base_dir = os.path.dirname(
        os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    )
    dir_path = os.path.join(base_dir, "logs", ano)
    file_path = os.path.join(dir_path, f"logs{tipo_txt}{mes}.txt")

    try:
        os.makedirs(dir_path, exist_ok=True)

        conteudo_existente = ""
        if os.path.exists(file_path):
            with open(file_path, "r", encoding="utf-8") as f:
                conteudo_existente = f.read()

        with open(file_path, "w", encoding="utf-8") as f:
            f.write(linha + conteudo_existente)
    except OSError as e:
        print(f"[Logger] Erro ao escrever .txt: {e}")

    # BD
    try:
        Database.inserir_log(tipo_bd, mensagem)
    except Exception as e:
        print(f"[Logger] Erro ao inserir log na BD: {e}")


# Logs de Utilizador

def log_utilizador(mensagem: str):
    _escrever_log("Utilizador", "utilizador", mensagem)

def log_registo(nome: str):
    log_utilizador(f"Novo utilizador registado com o nome: '{nome}'.")

def log_login(nome: str):
    log_utilizador(f"Utilizador '{nome}' efetuou login na sua conta.")

def log_logout(nome: str):
    log_utilizador(f"Utilizador '{nome}' deu logout da sua conta.")

def log_app_iniciada():
    log_utilizador("Aplicação iniciada.")

def log_app_terminada():
    log_utilizador("Aplicação terminada.")


# Logs de Cliente

def log_cliente(mensagem: str):
    _escrever_log("Clientes", "cliente", mensagem)

def log_cliente_criado(nome: str):
    log_cliente(f"Cliente '{nome}' criado.")

def log_cliente_apagado(nome: str):
    log_cliente(f"Cliente '{nome}' apagado.")

def log_falta_adicionada(nome: str):
    log_cliente(f"Falta adicionada ao Cliente '{nome}'.")

def log_falta_retirada(nome: str):
    log_cliente(f"Falta retirada ao Cliente '{nome}'.")

def log_nome_alterado(nome_antigo: str, nome_novo: str):
    log_cliente(f"Nome do Cliente '{nome_antigo}' alterado para '{nome_novo}'.")

def log_numero_alterado(nome: str, numero_novo: str):
    log_cliente(f"Numero do Cliente '{nome}' alterado para '{numero_novo}'.")

def log_tipo_semanal_para_normal(nome: str):
    log_cliente(f"Cliente '{nome}' alterado de SEMANAL para NORMAL.")

def log_tipo_normal_para_semanal(nome: str, dia: str, hora: str):
    log_cliente(
        f"Cliente '{nome}' alterado de NORMAL para SEMANAL, "
        f"horário de '{dia}' às '{hora}'."
    )

def log_horario_semanal_alterado(nome: str, dia: str, hora: str):
    log_cliente(
        f"Horas do Cliente SEMANAL '{nome}' alteradas para '{dia}' às '{hora}'."
    )


# Logs de Marcação

def log_marcacao(mensagem: str):
    _escrever_log("Marcacoes", "marcacao", mensagem)

def log_marcacao_criada(marcacao: Marcacao):
    cliente = marcacao.get_cliente()
    nome = cliente.get_nome() if cliente is not None else "N/A"
    data_hora = marcacao.get_data_hora()
    data_hora_str = data_hora.strftime("%d/%m/%Y %H:%M") if data_hora is not None else "N/A"
    log_marcacao(f"Marcação criada para '{nome}' em {data_hora_str}.")

def log_marcacao_apagada(marcacao: Marcacao):
    cliente = marcacao.get_cliente()
    nome = cliente.get_nome() if cliente is not None else "N/A"
    data_hora = marcacao.get_data_hora()
    data_hora_str = data_hora.strftime("%d/%m/%Y %H:%M") if data_hora is not None else "N/A"
    log_marcacao(f"Marcação de '{nome}' em {data_hora_str} apagada.")

def log_marcacao_obs_alterada(marcacao: Marcacao, obs_nova: str):
    cliente = marcacao.get_cliente()
    nome = cliente.get_nome() if cliente is not None else "N/A"
    data_hora = marcacao.get_data_hora()
    data_hora_str = data_hora.strftime("%d/%m/%Y %H:%M") if data_hora is not None else "N/A"
    log_marcacao(
        f"Observações da marcação de '{nome}' em {data_hora_str} "
        f"alteradas para '{obs_nova}'."
    )

def log_marcacao_data_hora_alterada(marcacao: Marcacao,
                                    data_hora_antiga: str,
                                    data_hora_nova: str):
    cliente = marcacao.get_cliente()
    nome = cliente.get_nome() if cliente is not None else "N/A"
    log_marcacao(
        f"Data/hora da marcação de '{nome}' alterada de "
        f"{data_hora_antiga} para {data_hora_nova}."
    )

def log_marcacao_falta(marcacao: Marcacao):
    cliente = marcacao.get_cliente()
    nome = cliente.get_nome() if cliente is not None else "N/A"
    data_hora = marcacao.get_data_hora()
    data_hora_str = data_hora.strftime("%d/%m/%Y %H:%M") if data_hora is not None else "N/A"
    log_marcacao(f"O cliente '{nome}' faltou a marcacao na data '{data_hora_str}'.")


# Logs de Pendentes

def log_pendente(mensagem: str):
    _escrever_log("Pendentes", "pendente", mensagem)

def log_pendente_adicionado(pendente: Pendente):
    log_pendente(f"Cliente Pendente adicionado: '{pendente.get_nome()}'.")

def log_pendente_removido(pendente: Pendente):
    log_pendente(f"Cliente Pendente removido: '{pendente.get_nome()}'.")

File: Java/src/utils/Logger.java
package utils;

import java.io.FileWriter;
import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import models.*;

public class Logger {
    private static String dataAtual() {
        return LocalDateTime.now().format(DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss"));
    }

    // Log generico Utilizador
    public static void logUtilizador(String mensagem) {
        LocalDateTime now = LocalDateTime.now();
        String data = dataAtual();
        String ano = now.format(DateTimeFormatter.ofPattern("yyyy"));
        String mes = now.format(DateTimeFormatter.ofPattern("MM"));
        String linha = "[" + data + "]" + mensagem + "\n";
        String dirPath = "logs/" + ano;
        String filePath = dirPath + "/logsUtilizador" + mes + ".json";

        try {
            // Garante que o diretório existe
            java.io.File dir = new java.io.File(dirPath);
            if (!dir.exists()) {
                dir.mkdirs();
            }

            // Ler o ficheiro se existir
            StringBuilder sb = new StringBuilder();
            java.io.File file = new java.io.File(filePath);
            if (file.exists()) {
                try (java.io.BufferedReader reader = new java.io.BufferedReader(new java.io.FileReader(file))) {
                    String l;
                    while ((l = reader.readLine()) != null) {
                        sb.append(l).append("\n");
                    }
                }
            }

            // Escrever o log
            try (FileWriter writer = new FileWriter(filePath, false)) {
                writer.write(linha);
                writer.write(sb.toString());
            }
        } catch (IOException e) {
            System.err.println("Erro ao escrever no ficheiro de log: " + e.getMessage());
        }
    }

    // Log Registo
    public static void logRegisto(String nome) {
        logUtilizador(" Novo utilizador registado com o nome: '" + nome + "'.");
    }

    // Log Login
    public static void logLogin(String nome) {
        logUtilizador(" Utilizador '" + nome + "' efetuou login na sua conta.");
    }

    // Log Logout
    public static void logLogout(String nome) {
        logUtilizador(" Utilizador '" + nome + "' deu logout da sua conta.");
    }

    // Log App Iniciada
    public static void logAppIniciada() {
        logUtilizador(" Aplicação iniciada.");
    }

    // Log App Terminada
    public static void logAppTerminada() {
        logUtilizador(" Aplicação terminada.");
    }

    // Log generico Cliete
    public static void logCliente(String mensagem) {
        LocalDateTime now = LocalDateTime.now();
        String data = dataAtual();
        String ano = now.format(DateTimeFormatter.ofPattern("yyyy"));
        String mes = now.format(DateTimeFormatter.ofPattern("MM"));
        String linha = "[" + data + "]" + mensagem + "\n";
        String dirPath = "logs/" + ano;
        String filePath = dirPath + "/logsClientes" + mes + ".json";

        try {
            // Garante que o diretório existe
            java.io.File dir = new java.io.File(dirPath);
            if (!dir.exists()) {
                dir.mkdirs();
            }

            // Ler o ficheiro se existir
            StringBuilder sb = new StringBuilder();
            java.io.File file = new java.io.File(filePath);
            if (file.exists()) {
                try (java.io.BufferedReader reader = new java.io.BufferedReader(new java.io.FileReader(file))) {
                    String l;
                    while ((l = reader.readLine()) != null) {
                        sb.append(l).append("\n");
                    }
                }
            }

            // Escrever o log
            try (FileWriter writer = new FileWriter(filePath, false)) {
                writer.write(linha);
                writer.write(sb.toString());
            }
        } catch (IOException e) {
            System.err.println("Erro ao escrever no ficheiro de log: " + e.getMessage());
        }
    }

    // Cliente criado
    public static void logClienteCriado(String nome) {
        logCliente("Cliente '" + nome + "' criado.");
    }

    // Cliente apagado
    public static void logClienteApagado(String nome) {
        logCliente("Cliente '" + nome + "' apagado.");
    }

    // Falta adicionada
    public static void logFaltaAdicionada(String nome) {
        logCliente("Falta adicionada ao Cliente '" + nome + "'.");
    }

    // Falta retirada
    public static void logFaltaRetirada(String nome) {
        logCliente("Falta retirada ao Cliente '" + nome + "'.");
    }

    // Nome alterado
    public static void logNomeAlterado(String nomeAntigo, String nomeNovo) {
        logCliente("Nome do Cliente '" + nomeAntigo + "' alterado para '" + nomeNovo + "'.");
    }

    // Número alterado
    public static void logNumeroAlterado(String nome, String numeroNovo) {
        logCliente("Numero do Cliente '" + nome + "' alterado para '" + numeroNovo + "'.");
    }

    // Tipo alterado de SEMANAL para NORMAL
    public static void logTipoSemanalParaNormal(String nome) {
        logCliente("Cliente '" + nome + "' alterado de SEMANAL para NORMAL.");
    }

    // Tipo alterado de NORMAL para SEMANAL
    public static void logTipoNormalParaSemanal(String nome, String dia, String hora) {
        logCliente(
                "Cliente '" + nome + "' alterado de NORMAL para SEMANAL, horário de '" + dia + "' às '" + hora + "'.");
    }

    // Horário alterado em SEMANAL
    public static void logHorarioSemanalAlterado(String nome, String dia, String hora) {
        logCliente("Horas do Cliente SEMANAL '" + nome + "' alteradas para '" + dia + "' às '" + hora + "'.");
    }

    // Log generico Marcacao
    public static void logMarcacao(String mensagem) {
        LocalDateTime now = LocalDateTime.now();
        String data = dataAtual();
        String ano = now.format(DateTimeFormatter.ofPattern("yyyy"));
        String mes = now.format(DateTimeFormatter.ofPattern("MM"));
        String linha = "[" + data + "] " + mensagem + "\n";
        String dirPath = "logs/" + ano;
        String filePath = dirPath + "/logsMarcacoes" + mes + ".json";

        try {
            // Garante que o diretório existe
            java.io.File dir = new java.io.File(dirPath);
            if (!dir.exists()) {
                dir.mkdirs();
            }

            // Ler o ficheiro se existir
            StringBuilder sb = new StringBuilder();
            java.io.File file = new java.io.File(filePath);
            if (file.exists()) {
                try (java.io.BufferedReader reader = new java.io.BufferedReader(new java.io.FileReader(file))) {
                    String l;
                    while ((l = reader.readLine()) != null) {
                        sb.append(l).append("\n");
                    }
                }
            }

            // Escrever o log
            try (FileWriter writer = new FileWriter(filePath, false)) {
                writer.write(linha);
                writer.write(sb.toString());
            }
        } catch (IOException e) {
            System.err.println("Erro ao escrever no ficheiro de log: " + e.getMessage());
        }
    }

    // Log de marcação criada
    public static void logMarcacaoCriada(Marcacao marcacao) {
        String nome = marcacao.getCliente() != null ? marcacao.getCliente().getNome() : "N/A";
        String dataHora = marcacao.getDataHora().format(DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm"));
        logMarcacao("Marcação criada para '" + nome + "' em " + dataHora + ".");
    }

    // Log de marcação apagada
    public static void logMarcacaoApagada(Marcacao marcacao) {
        String nome = marcacao.getCliente() != null ? marcacao.getCliente().getNome() : "N/A";
        String dataHora = marcacao.getDataHora().format(DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm"));
        logMarcacao("Marcação de '" + nome + "' em " + dataHora + " apagada.");
    }

    // Log de alteração de observações
    public static void logMarcacaoObsAlterada(Marcacao marcacao, String obsNova) {
        String nome = marcacao.getCliente() != null ? marcacao.getCliente().getNome() : "N/A";
        String dataHora = marcacao.getDataHora().format(DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm"));
        logMarcacao(
                "Observações da marcação de '" + nome + "' em " + dataHora + " alteradas" + " para '" + obsNova + "'.");
    }

    // Log de alteração de data/hora
    public static void logMarcacaoDataHoraAlterada(Marcacao marcacao, String dataHoraAntiga, String dataHoraNova) {
        String nome = marcacao.getCliente() != null ? marcacao.getCliente().getNome() : "N/A";
        logMarcacao("Data/hora da marcação de '" + nome + "' alterada de " + dataHoraAntiga + " para " + dataHoraNova
                + ".");
    }

    // Log de Falta à Marcacao
    public static void logMarcacaoFalta(Marcacao marcacao) {
        String nome = marcacao.getCliente() != null ? marcacao.getCliente().getNome() : "N/A";
        String dataHora = marcacao.getDataHora().format(DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm"));
        logMarcacao("O cliente '" + nome + "' faltou a marcacao na data '" + dataHora + "'.");
    }

    // Log genérico Pendente
    public static void logPendente(String mensagem) {
        LocalDateTime now = LocalDateTime.now();
        String data = dataAtual();
        String ano = now.format(DateTimeFormatter.ofPattern("yyyy"));
        String mes = now.format(DateTimeFormatter.ofPattern("MM"));
        String linha = "[" + data + "]" + mensagem + "\n";
        String dirPath = "logs/" + ano;
        String filePath = dirPath + "/logsPendentes" + mes + ".json";

        try {
            // Garante que o diretório existe
            java.io.File dir = new java.io.File(dirPath);
            if (!dir.exists()) {
                dir.mkdirs();
            }

            // Ler o ficheiro se existir
            StringBuilder sb = new StringBuilder();
            java.io.File file = new java.io.File(filePath);
            if (file.exists()) {
                try (java.io.BufferedReader reader = new java.io.BufferedReader(new java.io.FileReader(file))) {
                    String l;
                    while ((l = reader.readLine()) != null) {
                        sb.append(l).append("\n");
                    }
                }
            }

            // Escrever o log
            try (FileWriter writer = new FileWriter(filePath, false)) {
                writer.write(linha);
                writer.write(sb.toString());
            }
        } catch (IOException e) {
            System.err.println("Erro ao escrever no ficheiro de log: " + e.getMessage());
        }
    }

    // Log pendente adicionado
    public static void logPendenteAdicionado(Pendente pendente) {
        logPendente("Cliente Pendente adicionado: '" + pendente.getNome() + "'.");
    }

    // Log pendente removido
    public static void logPendenteRemovido(Pendente pendente) {
        logPendente("Cliente Pendente removido: '" + pendente.getNome() + "'.");
    }
}
